### ISSUE TITLE
Incorrect mixed type with Amphp

### ISSUE THREAD

**Issue by enumag** [vimeo/psalm#3418]
I'm using dev-master for this with your latest fixes. This bug still exists. Here is a reproducer:

```php
<?php declare(strict_types = 1);

namespace App;

use Amp\Promise;
use DateTime;
use Generator;
use function Amp\call;

abstract class Z
{
	/**
	 * @return Promise<DateTime>
	 */
	public function test(): Promise
	{
		return call(
			function (): Generator {
				$date = yield $this->load();

				return clone $date;
			}
		);
	}

	/**
	 * @return Promise<DateTime>
	 */
	abstract protected function load(): Promise;
}
```

This code reports:

```
ERROR: MixedClone - src/Z.php:21:12 - Cannot clone mixed (see https://psalm.dev/227)
                                return clone $date;
```

Which is a false-positive. From the return annotation of the load method it's clear that `$date` is of type `DateTime` and not mixed so it can be cloned just fine.

**Comment by EvgeniiR** [comment]
> From the return annotation of the load method it's clear that $date is of type DateTime and not mixed so it can be cloned just fine.

yield here doesn't necessarily mean that $date have type of DateTime, because of how generators work. 

Psalm in your situation just doesn't know the type of $date variable ( https://psalm.dev/r/7f3755cfa4 )

**Comment by enumag** [comment]
Afaik some support for Amphp was added recently which means Psalm is supposed to understand this.

**Comment by enumag** [comment]
ping @muglug

**Comment by enumag** [comment]
bump

**Comment by enumag** [comment]
@muglug How does this `@psalm-yield` annotation work? I'm quite interested to learn what it is about but I can't find an explanation anywhere. https://psalm.dev/docs/annotating_code/supported_annotations/
